### ISSUE TITLE
Multiple full-image bounding boxes receiving poor confidence scores

### ISSUE THREAD

**Issue by Transigent** [ultralytics/yolov5#6903]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I have been training a dataset of around 25k images for a while now, and gradually retraining for edge cases. In general the results have been really fantastic.
![Metrics](https://user-images.githubusercontent.com/18422245/157234604-ad5af22b-f283-443a-ac0d-5a24611a29bc.png)
![F1_curve_1](https://user-images.githubusercontent.com/18422245/157244496-a537e54b-b86c-4a45-a470-56e52cadb0fe.jpg)
![P_curve_1](https://user-images.githubusercontent.com/18422245/157234634-621b3f6d-fcbe-4359-b000-e461f2169e33.jpg)
![PR_curve_1](https://user-images.githubusercontent.com/18422245/157234647-6735b6e2-e875-41f3-b559-fe725c413dcb.jpg)
![R_curve_1](https://user-images.githubusercontent.com/18422245/157234652-f8e26ebf-e162-4317-a222-59fc2fb3719f.jpg)

But I have come across a situation that I can’t resolve because I’m not sure where the problem lies.  
I have annotated images with annotations as shown, and during inference the model works really well, the detected objects have confidence generally above 0.9. :

![RyanReynolds1sml](https://user-images.githubusercontent.com/18422245/157232988-50fa9ca6-30ed-4c0c-b898-a8cd2a900d62.jpg)
![Car1a](https://user-images.githubusercontent.com/18422245/157234905-6503c588-3e80-4285-9ee0-f0c5ecab219c.jpg)

In this case a Face and a Person, or a Car and a Wheel. They have no problem with being nested. 
 
But there are often other images like these, where class Person and class Face annotations are essentially overlapping (both classes are clearly valid, there is a face and its also a person). During inference I get very poor confidence scores for the detected objects of each class, often around 0.5 or lower , which puts them very close to the confidence threshold and in danger of being missed entirely.
![RyanReynolds2](https://user-images.githubusercontent.com/18422245/157233310-71159ca1-5848-46f5-beec-3322244df802.jpg)
![Car2](https://user-images.githubusercontent.com/18422245/157233318-72e530e9-3ca7-406b-9ea8-f065ec2f5d46.jpg)

I thought I would try and see what happens if I take an image that is exhibiting this problem during inference, then annotate it with both bounding boxes basically set to full image size, then copy it plus its annotation file say 10 times. I would then add all 10 back into the training set. To my way of thinking, if there were multiple of the same image with the same annotations, it would accelerate gradient descent and correct the issue strongly for at least that one image. 

After retraining using the dataset with the additional copies of the image, I found no improvement, I was still getting low confidence scores for that one image, for both annotations. This makes me wonder if there is something in the training code that removes one or the other if there are two closely overlapping annotations during training. What I would expect to see is both classes detected with very high confidence scores for (at least) that one image. 

I did read some other posts about multi_label NMS but they seem to be related to inferencing and I feel like this problem is more something to do with training. 

Any thoughts would be super welcome.

**Comment by glenn-jocher** [comment]
@Transigent 👋 Hello! Thanks for asking about improving YOLOv5 🚀 training results. Low confidences are an indicator of insufficient training, and as you can see in your results plots your model has not started overfitting yet, which is another indicator of lack of sufficient training time.

Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥ 1500 images per class recommended
- **Instances per class.** ≥ 10000 instances (labeled objects) per class recommended
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total). No labels are required for background images.

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

- **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. [VOC](https://github.com/ultralytics/yolov5/blob/master/data/VOC.yaml), [VisDrone](https://github.com/ultralytics/yolov5/blob/master/data/VisDrone.yaml), [GlobalWheat](https://github.com/ultralytics/yolov5/blob/master/data/GlobalWheat2020.yaml)). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
                                             custom_pretrained.pt
```
- **Start from Scratch.** Recommended for large datasets (i.e. [COCO](https://github.com/ultralytics/yolov5/blob/master/data/coco.yaml), [Objects365](https://github.com/ultralytics/yolov5/blob/master/data/Objects365.yaml), [OIv6](https://storage.googleapis.com/openimages/web/index.html)). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
```bash
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
                                                      yolov5m.yaml
                                                      yolov5l.yaml
                                                      yolov5x.yaml
```


## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyp.scratch.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains: http://karpathy.github.io/2019/04/25/recipe/

Good luck 🍀 and let us know if you have any other questions!

**Comment by Transigent** [comment]
Thanks for the reply @glenn-jocher , I've seen this article many times and have been using its guidelines since I started training. I have trained to 300 epochs before but I find no improvement in this specific issue (or generally) past about 150 epochs. 

I don't know if I described the issue clearly. I get generally excellent results with high confidence values across the board, except in the case described, when the images in the training set have overlapping, full image annotation rectangles. In an above example, a wheel and a car are both present in the zoomed image and should be detected, and to that end there is an annotation rectangle for each class in the training image that is essentially the size of the full image. In that specific case I get poor confidences for either class when inference is done. 

If however the wheels are smaller with regard to the size of the car, as in the full car image, confidences are above 0.9 as expected. 

I wonder if there is some logic during training that discards one or more annotations for differing classes that overlap each other and are of nearly the exact size and coverage.

**Comment by glenn-jocher** [comment]
@Transigent no there's nothing preventing overlapping boxes. 

In general full-image detection is not an intended use case though, as this is just classification with unneeded extra complexities for box regression and objectness, and would be better handled by a classification model like resnet.

**Comment by Transigent** [comment]
OK thanks anyway. I'll keep investigating.

**Comment by sarpx** [comment]
@glenn-jocher I am having the same problem. Objects are captured well in distant pictures, but the results are poor in close-ups. I also decided not to include close-ups in training, but this is not something I want.

**Comment by glenn-jocher** [comment]
@sarpx training and deployment space should be similar for best results.
![Screenshot 2021-08-04 at 13 15 05](https://user-images.githubusercontent.com/26833433/159535658-4f0a8da6-4f8c-4ca0-995f-81559c82f36d.jpg)

**Comment by Transigent** [comment]
Its interesting that I'm not the only one with this problem.

I have had to go through my training dataset and for any images that contain nearly-full-size overlapping annotation boxes, I have had to remove the least specific annotation, for example the car class and leave the wheel class there. Then later after inference I have to extrapolate up i.e. a wheel was detected so there implicitly has to be a car. 

This sucks heartily, but If I don't do this, there are cases where both the wheel and the car bounding boxes are close to the detection threshold and I am in danger of one or NEITHER being detected, especially with any confidence. After training with just the one annotation rectangle (wheel) its detection confidence is in the 0.95 range again.

It's like the two classes are not independent, and the presence of one class box with similar dimensions is affecting the performance of the other, and vice versa. I can't yet tell if the poor object confidence is baked into the model during training, or if it's something going on during inference. It only happens with closeups but I don't get to choose how 'zoomed' the inferenced images are.
